content/blog/2025-04-27-valkey-modules-rust-sdk-updates.md

```diff
@@ -0,0 +1,272 @@
+
+```rust
+fn my_client_cmd(ctx: &Context, _args: Vec<ValkeyString>) -> ValkeyResult {
+    let client_id = ctx.get_client_id();
+    let username = ctx.get_client_username_by_id(client_id);
+    Ok(ValkeyValue::from(username.to_string()))
+}
+valkey_module! {
+    ...
+    commands: [
+        ["my_client_cmd", my_client_cmd, "", 0, 0, 0],
+    ]
+}
+```
+
+## Auth callback
+
+Valkey 7.2 introduced support for callbacks after authentication.
+Now it can be leveraged via modules.
+
+```rust
+fn auth_callback(
+    ctx: &Context,
+    username: ValkeyString,
+    _password: ValkeyString,
+) -> Result<c_int, ValkeyError> {
+    if ctx.authenticate_client_with_acl_user(&username) == Status::Ok {
+        // can be combined with ctx.get_client_ip to control access by IP address
+        let _client_ip = ctx.get_client_ip()?;
+        ...
+        return Ok(AUTH_HANDLED);
+    }
+    Ok(AUTH_NOT_HANDLED)
+}
+valkey_module! {
+    ...
+    auth: [auth_callback],
+}
+```
+
+## Preload validation
+
+`valkey_module!` macro already provided `init` callback that can be used to execute custom code during module load.
+However, `init` executes at the very end of module load after new commands and data types are created.
+That can be use useful but what if we wanted to stop module load before any of that happens?
+For example, we might need to restrict module to be loaded only on specific version of Valkey.
+That's where the optional `preload` can be used.
+
+```rust
+fn preload_fn(ctx: &Context, _args: &[ValkeyString]) -> Status {
+    let _version = ctx.get_server_version().unwrap();
+    // respond with Status::Ok or Status::Err to prevent loading
+    Status::Ok
+}
+valkey_module! {
+    ...
+    preload: preload_fn,
+    commands: [],
+}
+```
+
+## Filters
+
+To execute custom code before specific Valkey commands we can use command filters.  This can be leveraged to replace default commands with custom comands or modify arguments.
+Thanks to the abstractions provided by the SDK we simply need to create a Rust function and register it in the `valkey_module!` macro.
+Note of caution - since filters are executed before every command this code needs to be optimized.
+
+```rust
+fn my_filter_fn(_ctx: *mut RedisModuleCommandFilterCtx) {
+    let cf_ctx = CommandFilterCtx::new(ctx);
+    // check the number of arguments
+    if cf_ctx.args_count() != 3 {
+        return;
+    }
+    // get which command is being executed
+    let _cmd = cf_ctx.cmd_get_try_as_str().unwrap();
+    // grab various args passed to the command
+    let _all_args = cf_ctx.get_all_args_wo_cmd();
+    // replace command
+    cf_ctx.arg_replace(0, "custom_cmd");
+}
+valkey_module! {
+    ...
+    filters: [
+        [my_filter_fn, VALKEYMODULE_CMDFILTER_NOSELF],
+    ]
+}
+```
+
+## New event handlers
+
+SDK now supports more server events.  We can use this to execute our own code on client connect/disconnect, server shutdown or specific key events.
```

Comment:
definite article: '**The** SDK now supports...'

Reply:
updated